Content.Server/Lightning/LightningSystem.cs

```diff
@@ -69,70 +49,54 @@ private void OnCollide(EntityUid uid, LightningComponent component, ref StartCol
     /// <param name="lightningPrototype">The prototype for the lightning to be created</param>
     public void ShootLightning(EntityUid user, EntityUid target, string lightningPrototype = "Lightning")
     {
+        if (Deleted(user) || Deleted(target))
+            return;
+
         var spriteState = LightningRandomizer();
         _beam.TryCreateBeam(user, target, lightningPrototype, spriteState);
+
+        var ev = new HitByLightningEvent(user, target);
+        RaiseLocalEvent(target, ref ev, true);
     }
 
     /// <summary>
-    /// Looks for a target to arc to in all 8 directions, adds the closest to a local dictionary and picks at random
+    /// Looks for objects with a LightningTarget component in the radius, prioritizes them, and hits the highest priority targets with lightning.
     /// </summary>
-    /// <param name="component"></param>
-    /// <param name="target"></param>
-    /// <param name="controllerEntity"></param>
-    private void Arc(LightningComponent component, EntityUid target, EntityUid controllerEntity)
+    /// <param name="user">Where the lightning fires from</param>
+    /// <param name="range">Targets selection radius</param>
+    /// <param name="boltCount">Number of lightning bolts</param>
+    /// <param name="lightningPrototype">The prototype for the lightning to be created</param>
+    /// <param name="arcDepth">how many times to recursively fire lightning bolts from the target points of the first shot.</param>
+    public void ShootRandomLightnings(EntityUid user, float range, int boltCount, string lightningPrototype = "Lightning", int arcDepth = 0)
     {
-        if (!TryComp<BeamComponent>(controllerEntity, out var controller))
-            return;
-
-        var targetXForm = Transform(target);
-        var directions = Enum.GetValues<Direction>().Length;
+        //To Do: add support to different priority target tablem for different lightning types
+        var targets = _lookup.GetComponentsInRange<LightningTargetComponent>(Transform(user).Coordinates, range);
+        var sortedTargets = targets
+            .OrderByDescending(target => target.Priority)
+            .ThenBy(_ => _random.Next())
+            .ToList();
```

Comment:
This is expensive and should be optimised, nested LINQ calls should pretty much never be used in gamedev as it allocates every call.

Reply:
Check this
```csharp
var targets = _lookup.GetComponentsInRange<LightningTargetComponent>(Transform(user).Coordinates, range).ToList(); //To Do: remove hardcode component
        _random.Shuffle(targets);
        targets.Sort((x, y) => y.Priority.CompareTo(x.Priority));
```